include/common/logging.h

```diff
@@ -0,0 +1,15 @@
+#ifndef _ROSE_C_LOGGING_H
+#define _ROSE_C_LOGGING_H
+
+#include "common/printf.h"
+
+#define LOG(msg)        do {                        \
+    printf(__FILE__":%d: %s\r\n", __LINE__, msg);   \
+} while (0)
```

Comment:
What is the point of a `do {} while (0)` loop? Can't you just do printf?

Reply:
Remember that macros just do a copy paste. This requires some preventive measures for safe use. Also, ideally we want to model a macro to work like any function call.

why `{...}`:
* The braces prevent accidental 
* Putting the macro into curly braces allows adding arbitrarily many lines of code while still counting as one expression. Without it, multi-expression macros couldn't be used in places that expect only one macro, e.g.:
```
if (cond) LOG("cond is true");
```

why `do {...} while (0)`:
* It forces the use of a semicolon at the end.
  * Why would I want that?
    * Your IDE doesn't complain
    * it makes the macro look more like code and less like macro (-> code looks more consistent)

why _not_ `{...}; false;`?
* Putting another expression at the end can lead to unwanted side effects (in this case, the whole expression evaluates to `false`)
* It's counted as two expressions again. Of course we don't have to care about that, and it usually works, but there may be cases where it suddenly doesn't.

Maybe I find something cleaner than `do {...} while (0)`, but for now it works without any unintended side effects

Reply:
Note to the last point: I haven't found any expression you could put after a `{...};` block to force the semicolon.
* Any value would change what the macro returns
* Putting something like `void` there works but gives a compiler warning about an unused type
* You could write a function `void _(){}` and then wrap the macro into a `{...}; _()` but that adds a weird unused global function (plus, it doesn't add to the legibility of the macro)
* Writing `{...}; if (0)` also works, but at that point, you might as well stick to the one expression variant of `do {...} while (0)` (plus that might also confuse your IDE)
*

Reply:
Hm, honestly that feels so hacky that I don't see how that could be intended :sweat_smile: 

I pressume that just leaving away the semicolon behind the printf line would make your IDE complain too then?

Reply:
That would only work without the braces.
Like C, the preprocessor is a very basic tool. Either we build abstractions around it to make it look like C or we use it with the fact that it's a macro in mind.

In this case, it's pretty clear what the macro does. You won't ever need to look at it again because you know how it works. Therefore, it fulfils its purpose of abstracting features of the preprocessor to look like normal C code

Reply:
Hm, I see your points, though in this particular case at least I don't think it makes so much sense to add all this noise into an effective oneliner just because of premature optimization. But in the end that's just style. I don't see what would be added to this macro down the line, maybe you have ideas.

Also, apparently you can also just use parentheses instead of do {} while(0) https://www.geeksforgeeks.org/multiline-macros-in-c/

Reply:
> Also, apparently you can also just use parentheses instead of do {} while(0)
As I explained in my third argument, whilst possible, it doesn't require a semicolon at the end which in turn confuses IDEs

Reply:
> I don't see what would be added to this macro down the line, maybe you have ideas.

Right now, the logging doesn't support formatting like `printf` does. We will get annoyed at some point because we can't have a nice one-liner to log variables. Implementing this requires a far more complex setup, though.